pydantic/_internal/_core_utils.py

```diff
@@ -66,6 +67,7 @@ def is_list_like_schema_with_items_schema(
     return schema['type'] in _LIST_LIKE_SCHEMA_WITH_ITEMS_TYPES
 
 
+@lru_cache
```

Comment:
Should it be some weak cache like in generics side to avoid (dynamic cough cough) models being leaked.

Reply:
Yeah, I currently went with the default maxsize here (128) to avoid memory leaks, but this can still hold references to types/classes that should be gc'd, and with a small application where you have less than 128 referenceable types, this will never reach the limit and thus keep a reference to all of them.

It's a bit annoying that the `functools` cache utilities are limited in that aspect. It would be great if we could define our own cache key (currently hash) and customize the internal cache implementation (currently a dict). The `cachetools` 3rd party lib states [it supports](https://cachetools.readthedocs.io/en/latest/#cachetools.cached) using a `WeakValueDictionary`.

On the K8S file, with the cache impl, `get_type_ref` takes 30ms/~7s, without it takes 100ms/~7s, so I think we can live without it. Alternatively, we could a class level cached property on Pydantic models so that we only compute the reference once, I can take a look.

Reply:
It could maybe simple as:

```python
class _TypeRefCache(WeakKeyDictionary):
    def __setitem__(self, key: type[Any], value: Any) -> None:
        size = len(self)
        if size >= 1000:
            for remove_key in [*self.keys()][: size // 10]:  # Remove 10% of the cache (FIFO)
                del self[remove_key]
        super().__setitem__(key, value)


_TYPE_REF_CACHE = _TypeRefCache()


def get_type_ref(type_: type[Any], args_override: tuple[type[Any], ...] | None = None) -> str:
    """...."""
    if isinstance(type_, type) and not args_override:
        if (cached := _TYPE_REF_CACHE.get(type_)) is not None:
            return cached
        type_ref = _get_type_ref(type_, args_override)
        _TYPE_REF_CACHE[type_] = type_ref
        return type_ref
    else:
        return _get_type_ref(type_, args_override)


def _get_type_ref(type_: type[Any], args_override: tuple[type[Any], ...] | None = None) -> str:
    ... rest of stuff
```

Reply:
A better API would be for each type to compute the ref, as currently `get_type_ref` is doing useless work as it needs to account for type alias types, dataclasses, etc. Instead, we should be able to decouple everything from this function. Postponing for now.